src/lib/agent.ts

```diff
@@ -62,13 +69,71 @@ export class Agent {
     eventbus.on("didGenerated", callback)
   }
 
+  private handleDiscoverFeatures(message: IMessage) {
+    const regexEscape = (s: string) => s.replace(/([.*+?$^=!:{}()|\[\]\/\\])/g, "\\$1");
+    const createRegex = (query: string) => (new RegExp(`^${query.split("*").map(regexEscape).join(".*")}$`));
+    let protocolResponse: object[] = [];
+
+    // Loop through all queries, then all implemented protocols and build up a
+    // list of supported protocols that match the user's request
+    for(let query of message.body.queries) {
+
+      // Rudimentary implementation, ignoring all except protocol requests
+      if(query["feature-type"] != "protocol")
+        continue
+
+      for(let protocol of IMPLEMENTED_MESSAGE_TYPES) {
+        if(createRegex(query["match"]).test(protocol)) {
+          protocolResponse.push({
+            "feature-type": "protocol",
+            "id": protocol,
+          })
+        }
+      }
+    }
```

Comment:
Not really important but this is a good case for the `filter` method :slightly_smiling_face:

Reply:
Yep, didn't think of that since I don't use JS all too much. I'll have to look into how best to do this later :thinking: